pkg/controller.v1beta1/trial/trial_controller_util.go

```diff
@@ -171,34 +165,51 @@ func isJobSucceeded(jobCondition *commonv1.JobCondition) bool {
 	return false
 }
 
-func getBestObjectiveMetricValue(metricLogs []*api_pb.MetricLog, objectiveType commonv1beta1.ObjectiveType) *string {
-	metricLogSize := len(metricLogs)
-	if metricLogSize == 0 {
-		return nil
-	}
-
-	bestObjectiveValue, err := strconv.ParseFloat(metricLogs[0].Metric.Value, 64)
-	if err != nil {
-		// If metrics are string values return the latest value
-		return &metricLogs[len(metricLogs)-1].Metric.Value
+func getMetrics(metricLogs []*api_pb.MetricLog, strategies map[string]commonv1beta1.MetricStrategy) (*commonv1beta1.Observation, error) {
+	metrics := make(map[string]*commonv1beta1.Metric)
+	timestamps := make(map[string]*time.Time)
+	for name := range strategies {
+		timestamps[name] = nil
+		metrics[name] = &commonv1beta1.Metric{
+			Name:   name,
+			Min:    math.NaN(),
+			Max:    math.NaN(),
+			Latest: "",
+		}
 	}
-	bestIndex := 0
 
-	for idx, metricLog := range metricLogs[1:] {
-		objectiveMetricValue, _ := strconv.ParseFloat(metricLog.Metric.Value, 64)
-		if objectiveType == commonv1beta1.ObjectiveTypeMinimize {
-			if objectiveMetricValue < bestObjectiveValue {
-				bestObjectiveValue = objectiveMetricValue
-				bestIndex = idx + 1
+	for _, metricLog := range metricLogs {
+		metric, ok := metrics[metricLog.Metric.Name]
+		if !ok {
+			continue
+		}
+		strValue := metricLog.Metric.Value
+		floatValue, err := strconv.ParseFloat(strValue, 64)
+		if err == nil {
+			if math.IsNaN(metric.Min) || floatValue < metric.Min {
+				metric.Min = floatValue
 			}
-		} else if objectiveType == commonv1beta1.ObjectiveTypeMaximize {
-			if objectiveMetricValue > bestObjectiveValue {
-				bestObjectiveValue = objectiveMetricValue
-				bestIndex = idx + 1
+			if math.IsNaN(metric.Max) || floatValue > metric.Max {
+				metric.Max = floatValue
 			}
 		}
+		currentTime, err := time.Parse(time.RFC3339Nano, metricLog.TimeStamp)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse timestamps %s: %e", metricLog.TimeStamp, err)
+		}
+		timestamp, _ := timestamps[metricLog.Metric.Name]
+		if timestamp == nil || !timestamp.After(currentTime) {
```

Comment:
What if we have more than one metric with the same timestamp? 
Maybe we can extract Latest metric using latest element from the `metricLogs` for each metric name, what do you think?

Reply:
I agree. And I think, with current implementation, katib will use latest element if there exists multiple metricLogs with same timestamp. 
The condition `!timestamp.After(currentTime)` ensure latest metric value will update only if current timestamp is not earlier than historical latest timestamp. So, when they are equal, the updation will still be applied.

Reply:
Should we change it to `!currentTime.After(timestamp)` ? 
Since `timestamp` <=  `currentTime`, because `currentTime` is the new value from `metricLog` and `timestamp` historical value that we have recorded ?

Reply:
@sperlingxx Have you tried to test it?

Reply:
Hi @andreyvelich , I missed this comment sereval days ago.
 I think `!timestamp.After(currentTime)` represents `not ({timestamp} > {currentTime})`, which equals to ` {timestamp} <= {currentTime}`.
And there is a test case on extracting latest records if their values are same. Here is the [link](https://github.com/kubeflow/katib/pull/1140/files#diff-89fc68ee85fb96116fd4b809f044ea8bR256).

Reply:
@sperlingxx Got it. Thank you